AP_NavEKF3: correct GPS for position as it is recalled from buffer

rmackay9 opened the pull request:
This PR makes two small changes to when EKF3 corrects the GPS position and velocity for antenna position.  These changes have been broken out from the EKF active source PR https://github.com/ArduPilot/ardupilot/pull/14803 in order to reduce its complexity.

1. gpsDataDelayed is corrected immediately after it is recalled from the buffer instead of a few lines lower down.  The functional impact is that the delayed GPS data is always corrected even if it is not fused.  This is important for the larger PR because the GPS is more likely to be enabled but not used.  More broadly though this reduces the chance that uncorrected GPS data could be used for altitude or heading
2. CorrectGPSForAntennaOffset always corrects the velocity data instead of only when **fuseVelData** is true.

  - The advantage is that this allows the GPS velocities to be used (for yaw or whatever) even if they are not being fused into the EKF's velocity estimate.  Also **fuseVelData** can be true when we are fusing ExtNav velocities.
  - This also reduces the chance of errors by callers who are not aware of this hidden **fuseVelData** check within the function.  For example ResetVelocity() uses the method to correct **gpsDataNew**.
  - The danger is that some parts of the code rely on the GPS velocity values being zero as a way to know whether the GPS provides velocity instead of checking "frontend->_fusionModeGPS <= 1" (Note: 0 = use 3D velocity, 1 = use 2D velocity, 2 = use no velocity).  [realignYawGPS()](https://github.com/ArduPilot/ardupilot/blob/master/libraries/AP_NavEKF3/AP_NavEKF3_MagFusion.cpp#L139) is the only case I found where this **could** be a problem but I don't think it actually is because it checks for a velocity of 5m/s which is quite a large velocity to be caused by an attitude change.

This change has been tested in SITL:

- AHRS_EKF_TYPE=3, EK3_ENABLE=1, EK2_ENABLE=0 (to enabled EKF3)
- param set SIM_GPS_POS1_Z 5 (to simulate GPS placed 5m below frame)
- param set GPS_POS1_Z 5 (to configure AP to know the GPS is placed 5m below frame)
- Loiter
- arm throttle
- rc 3 2000 (to climb to 10m)
- rc 3 1500 (to hover)
- rc 2 1200 (to fly foward quickly about 50m)
- rc 2 1500 (to level out)
- RTL (to complete the test)

Below are screeshots of the resulting "before" and "after" NED velocity innovations and we see they are very similar.  Perhaps "after" is slightly better but this is likely just "environment" factors because the tests were run manually so inevitably the vehicle reached a slightly different speed, etc.
![before-after](https://user-images.githubusercontent.com/1498098/96806339-c9c3f780-144e-11eb-9d94-b3e3841fe3ee.png)

These changes have also been tested on a real vehicle as part of the larger active EKF source PR https://github.com/ArduPilot/ardupilot/pull/14803

tridge commented:
this looks right, but would we be better off having a flags byte in the gps_elements structure, marking if it had been corrected? The tricky code paths are the rarely used reset ones, and those are hard to test for properly

rmackay9 commented:
I've added the flag and re-tested to show that the results are as expected
![after-with-corrected-flag](https://user-images.githubusercontent.com/1498098/96868128-931acb00-14a8-11eb-91e2-f0e1af45c795.png)

During development I also added some debug to be sure the flag was being set properly as data was added and taken out of the buffer.  The debug just shows that immediately after the GPS data was recalled from the buffer the flag was false.  A couple of lines later after it had been Corrected it was true.
![image](https://user-images.githubusercontent.com/1498098/96868190-a9288b80-14a8-11eb-8e19-5637ca4d5372.png)

rmackay9 commented:
A couple of addition notes:

- There are two places where we could remove a "gps_corrected" temporary variable and instead simply directly update gpsDataNew.  I'm happy to make this change.
- During simple testing I didn't find any double corrections happening
- NavEKF3_core::calcGpsGoodToAlign() is using uncorrected velocity data before takeoff when it "Check that the vertical GPS vertical velocity is reasonable after noise filtering".  I suspect this is not a big problem but this was found using the new flag.

tridge reviewed (approved):
looks good, thanks!
I just tested it with the key test from my point of view, which is setup dual GPS with large separation, then force GPS switch by changing GPS_AUTO_SWITCH. The copter doesn't go crazy (it used to before a previous fix)

rmackay9 commented:
Merged, thanks!
